src/io/parquet/geoparquet_metadata.rs

```diff
@@ -43,3 +48,87 @@ impl GeoParquetMetadata {
+    if geometry_types.iter().any(|t| t.contains(" Z")) {
+        return Err(GeoArrowError::General(
+            "3D coordinates not currently supported".to_string(),
+        ));
+    }
+
+    match geometry_types.len() {
+        0 => Ok(None),
+        1 => Ok(Some(match *geometry_types.iter().next().unwrap() {
+            "Point" => GeoDataType::Point(coord_type),
+            "LineString" => GeoDataType::LineString(coord_type),
+            "Polygon" => GeoDataType::Polygon(coord_type),
+            "MultiPoint" => GeoDataType::MultiPoint(coord_type),
+            "MultiLineString" => GeoDataType::MultiLineString(coord_type),
+            "MultiPolygon" => GeoDataType::MultiPolygon(coord_type),
+            "GeometryCollection" => GeoDataType::GeometryCollection(coord_type),
+            _ => unreachable!(),
+        })),
+        2 => {
+            if geometry_types.contains("Point") && geometry_types.contains("MultiPoint") {
+                Ok(Some(GeoDataType::MultiPoint(coord_type)))
+            } else if geometry_types.contains("LineString")
+                && geometry_types.contains("MultiLineString")
+            {
+                Ok(Some(GeoDataType::MultiLineString(coord_type)))
+            } else if geometry_types.contains("Polygon") && geometry_types.contains("MultiPolygon")
+            {
+                Ok(Some(GeoDataType::MultiPolygon(coord_type)))
+            } else {
+                Ok(Some(GeoDataType::Mixed(coord_type)))
+            }
+        }
+        _ => Ok(Some(GeoDataType::Mixed(coord_type))),
+    }
+}
+
+fn parse_geoparquet_metadata(
+    metadata: &FileMetaData,
+    schema: &Schema,
+    coord_type: CoordType,
+) -> Result<(usize, Option<GeoDataType>)> {
+    let meta = GeoParquetMetadata::from_parquet_meta(metadata)?;
+    let column_meta = meta
+        .columns
+        .get(&meta.primary_column)
+        .ok_or(GeoArrowError::General(format!(
+            "Expected {} in GeoParquet column metadata",
+            &meta.primary_column
+        )))?;
+
+    let geometry_column_index = schema
+        .fields()
+        .iter()
+        .position(|field| field.name() == &meta.primary_column)
+        .unwrap();
+    let mut geometry_types = HashSet::with_capacity(column_meta.geometry_types.len());
+    column_meta.geometry_types.iter().for_each(|t| {
+        geometry_types.insert(t.as_str());
+    });
+    Ok((
+        geometry_column_index,
+        infer_geo_data_type(&geometry_types, coord_type)?,
+    ))
+}
+
+pub fn build_arrow_schema<T>(
+    builder: &ArrowReaderBuilder<T>,
+    coord_type: &CoordType,
+) -> (Arc<Schema>, usize, Option<GeoDataType>) {
+    let parquet_meta = builder.metadata();
+    let arrow_schema = builder.schema().clone();
+    let Ok((geometry_column_index, target_geo_data_type)) =
+        parse_geoparquet_metadata(parquet_meta.file_metadata(), &arrow_schema, *coord_type)
+    else {
+        panic!("Cannot parse geoparquet metadata");
+    };
```

Comment:
Probably should return a GeoArrowError here instead of panic? A little unsure if this let-else syntax is the idiomatic way of doing things...

Reply:
You should change the return type of this function to 
```rs
-> Result<(Arc<Schema>, usize, Option<GeoDataType>)> {
```
(note that this `Result` is pointing to `crate::error::Result`, which shadow's the standard-lib's `Result`. See https://github.com/geoarrow/geoarrow-rs/blob/eb94df3d998801be3c55128b4535fc9f58652b66/src/error.rs#L73

Then now that this function returns a `Result`, you can use the `?` operator to handle exceptions. So this can change to
```suggestion
    let (geometry_column_index, target_geo_data_type) =
        parse_geoparquet_metadata(parquet_meta.file_metadata(), &arrow_schema, *coord_type)?;
    Ok((arrow_schema, geometry_column_index, target_geo_data_type))
```
That is, we call the first line, unwrapping the result if it's an `Ok` or propagating the error upwards if it's an `Err`, and then we return an `Ok` with the data to return from this function

Reply:
Then anywhere that calls `build_arrow_schema` can use `?` to unwrap the error.

Reply:
Ah yes, returning `Result` makes more sense. Done in 97b59110465a5b72f74e72145d374f4e7d92ffd3.